tests/versioned/koa.tap.js

```diff
@@ -0,0 +1,219 @@
+'use strict'
+
+var tap = require('tap')
+var utils = require('@newrelic/test-utilities')
+var http = require('http')
+
+utils(tap)
+
+tap.test('Koa instrumentation', function(t) {
+  var helper = null
+  var app = null
+  var server = null
+
+  t.beforeEach(function(done) {
+    helper = utils.TestAgent.makeInstrumented()
+    helper.registerInstrumentation({
+      moduleName: 'koa',
+      type: 'web-framework',
+      onRequire: require('../../lib/instrumentation')
+    })
+    var Koa = require('koa')
+    app = new Koa()
+    done()
+  })
+
+  t.afterEach(function(done) {
+    server.close()
+    app = null
+    helper && helper.unload()
+    done()
+  })
+
+  t.test('produces transaction trace with multiple middleware', function(t) {
+    app.use(function one(ctx, next) {
+      return next()
+    })
+    app.use(function two(ctx, next) {
+      return next()
+    })
+    app.use(function three(ctx, next) {
+      next()
+    })
+
+    helper.agent.on('transactionFinished', function(tx) {
+      checkMiddlewareSegments(t, tx)
+    })
+
+    run()
+  })
+
+  t.test('correctly records actions interspersed among middleware', function(t) {
+    helper.agent.config.allow_all_headers = true
+    app.use(function one(ctx, next) {
+      helper.agent.tracer.createSegment('testSegment')
+      return next().then(function() {
+        helper.agent.tracer.createSegment('nestedSegment')
+        ctx.set('X-Response-Time', '1ms')
+      })
+    })
+    app.use(function two(ctx, next) {
+      return new Promise(function(resolve) {
+        setTimeout(resolve, 10)
+      })
+      .then(next)
+    })
+    app.use(function three(ctx) {
+      ctx.body = 'done'
+    })
+
+    helper.agent.on('transactionFinished', function(tx) {
+      var trace = tx.trace.root.children[0]
+      var attributes = tx.trace.attributes.attributes
+      t.ok(
+        attributes['response.headers.xResponseTime'],
+        'collected expected response header'
+      )
+      var mid1 = trace.children[0]
+      t.equal(
+        mid1.name,
+        'Nodejs/Middleware/Koa/one',
+        '1st middleware segment has expected name'
+      )
```

Comment:
You should use the `t.segments` or `t.exactSegments` tests instead of manually testing things. I believe the syntax is something like this:
```js
t.exactSegments(tx.trace.root, [{
  name: "Nodejs/Middleware/Koa/one",
  children: [{
    name: "Truncated/testSegment"
  }, {
    name: "Nodejs/Middleware/Koa/two"
    children: [...]
  }, {
    name: "Truncated/nestedSegment"
  }]
}])
```